### PR TITLE
fix: converted wizard to Typescript

### PR DESCRIPTION
**Pull request by kyegupov**
- [ ] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?
Makes the wizard a bit less magical.

#### Where should the reviewer start?
Wherever they can pour themselves a hard drink. This code needs some preparation.

### CODE REVIEW

**Comment by lili2311** [comment]
@kyegupov please re-base?

File: src/cli/commands/protect/wizard.ts [hunk 1]
```diff
@@ -249,7 +258,7 @@ function addProtectScripts(existingScripts, npmVersion, options) {
     return scripts;
   }
 
-  const npmVersionMajor = parseInt(npmVersion.split('.')[0]);
+  const npmVersionMajor = parseInt(npmVersion.split('.')[0], 10);
```
Comment (miiila) [inline src/cli/commands/protect/wizard.ts]: Very nice!

File: src/cli/commands/protect/wizard.ts [hunk 2]
```diff
@@ -295,18 +311,18 @@ function processAnswers(answers, policy, options) {
     }
 
     const answer = answers[key];
-    const res = {
+    const entry = {
       vulnId: answer.vuln.id,
       choice: answer.choice,
       from: answer.vuln.from.slice(1),
-    };
+    } as any;
```
Comment (miiila) [inline src/cli/commands/protect/wizard.ts]: Why any? Don't we know at least the basic interface?

File: src/cli/commands/protect/wizard.ts [hunk 3]
```diff
@@ -484,7 +500,7 @@ function processAnswers(answers, policy, options) {
 
         const lbl = 'Updating npm-shrinkwrap.json...';
         return spinner(lbl)
-          .then(npm.bind(null, 'shrinkwrap', null, live, cwd, null))
+          .then(npm.bind(null, 'shrinkwrap', null, live as any, cwd, null))
```
Comment (miiila) [inline src/cli/commands/protect/wizard.ts]: I hope `live` can be `boolean`, but definitely not sure...

File: src/cli/commands/protect/wizard.ts [hunk 4]
```diff
@@ -520,8 +536,8 @@ function processAnswers(answers, policy, options) {
       }
 
       return info.inspect(cwd, targetFile, options)
-        .then(spinner(lbl))
-        .then(snykMonitor.bind(null, cwd, meta))
+        .then(() => spinner(lbl))
```
Comment (miiila) [inline src/cli/commands/protect/wizard.ts]: Why is this needed? Just curious, I don't see the difference :-)

File: src/cli/commands/protect/wizard.ts [hunk 5]
```diff
@@ -125,11 +130,15 @@ function processWizardFlow(options) {
               // We need to have modules information for remediation. See Payload.modules
               options.traverseNodeModules = true;
 
-              return snyk.test(cwd, options).then((res) => {
+              return snyk.test(process.cwd(), options).then((oneOrManyRes) => {
+                if (oneOrManyRes[0]) {
```
[the remark this answers is not in the document]
Comment (kyegupov) [reply src/cli/commands/protect/wizard.ts]: Um, how exactly would such a test work?

File: src/cli/commands/protect/wizard.ts (hunk 2 above)
Comment (kyegupov) [reply src/cli/commands/protect/wizard.ts]: Well, it's analytics. We don't care that much (yet?).

File: src/cli/commands/protect/wizard.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (miiila) [reply src/cli/commands/protect/wizard.ts]: Stubing `snyk.test` to return an array and asserting that wizard fails with expecting error message?

File: src/cli/commands/protect/wizard.ts (hunk 3 above)
Comment (kyegupov) [reply src/cli/commands/protect/wizard.ts]: Got rid of bind(), converted npm.js to TS

File: src/cli/commands/protect/wizard.ts (hunk 4 above)
Comment (kyegupov) [reply src/cli/commands/protect/wizard.ts]: I have no idea how it worked before (it shouldn't have). Anyway, rewritten the spinner in Typescript.

File: src/cli/commands/protect/prompts.ts [hunk 6]
```diff
@@ -316,12 +322,11 @@ function getPatchPrompts(vulns, policy, options) {
 
   const prompts = generatePrompt(res, policy, 'p', options);
 
-
   return prompts;
 
 }
 
-function getIgnorePrompts(vulns, policy, options) {
+function getIgnorePrompts(vulns, policy, options?) {
```
[the remark this answers is not in the document]
Comment (kyegupov) [reply src/cli/commands/protect/prompts.ts]: Yup, looks like it. Fixed the parameter types.

File: src/cli/commands/protect/prompts.ts [hunk 7]
```diff
@@ -553,25 +600,25 @@ function generatePrompt(vulns, policy, prefix, options) {
...
         note = '   Note: ' + vuln.note;
       }
     }
 
-    const res = {
-      when: function (answers) {
-        let res = true;
+    const res: Prompt = {
+      when(answers) {
+        let res2 = true;
```
[the remark this answers is not in the document]
Comment (kyegupov) [reply src/cli/commands/protect/prompts.ts]: Me too. Speculatively changed to `haventUpgraded`

File: src/cli/commands/protect/prompts.ts [hunk 8]
```diff
@@ -772,8 +818,8 @@ function generatePrompt(vulns, policy, prefix, options) {
     // choice and whether it was supposed to be a default. If the user is
     // updating their policy options, then we select the choice they had
     // before, otherwise we select the default
-    res.default = (choices.map((choice, i) => {
-      return {i: i, default: choice.default};
+    res.default = (choices.map((choice, ii) => {
```
[the remark this answers is not in the document]
Comment (kyegupov) [reply src/cli/commands/protect/prompts.ts]: cIndex for choice index

File: src/cli/commands/protect/wizard.ts [hunk 9]
```diff
@@ -1,52 +1,57 @@
...
-const detect = require('../../../lib/detect');
-const plugins = require('../../../lib/plugins');
-const moduleInfo = require('../../../lib/module-info').ModuleInfo;
-const {MisconfiguredAuthInCI} = require('../../../lib/errors/misconfigured-auth-in-ci-error');
-const {MissingTargetFileError} = require('../../../lib/errors/missing-targetfile-error');
-const pm = require('../../../lib/package-managers');
-
-function wizard(options = {}) {
+Object.assign(wizard, {
+   processAnswers,
```
[the remark this answers is not in the document]
Comment (kyegupov) [reply src/cli/commands/protect/wizard.ts]: We'll introduce Prettier soon, I believe.

File: src/cli/commands/protect/wizard.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (kyegupov) [reply src/cli/commands/protect/wizard.ts]: How would that test work?
